src/sql/workbench/services/executionPlan/common/executionPlanService.ts

```diff
@@ -0,0 +1,130 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { localize } from 'vs/nls';
+import type * as azdata from 'azdata';
+import { IExecutionPlanService } from 'sql/workbench/services/executionPlan/common/interfaces';
+import { ICapabilitiesService } from 'sql/platform/capabilities/common/capabilitiesService';
+import { IQuickInputService, IQuickPickItem } from 'vs/platform/quickinput/common/quickInput';
+import { IExtensionService } from 'vs/workbench/services/extensions/common/extensions';
+import { Event, Emitter } from 'vs/base/common/event';
+
+interface ExecutionPlanProviderRegisteredEvent {
+	id: string,
+	provider: azdata.executionPlan.ExecutionPlanServiceProvider
+}
+export class ExecutionPlanService implements IExecutionPlanService {
+	private _providers: { [handle: string]: azdata.executionPlan.ExecutionPlanServiceProvider; } = Object.create(null);
+	private _onProviderRegister: Emitter<ExecutionPlanProviderRegisteredEvent> = new Emitter<ExecutionPlanProviderRegisteredEvent>();
+	private _providerRegisterEvent: Event<ExecutionPlanProviderRegisteredEvent>;
+	constructor(
+		@ICapabilitiesService private _capabilitiesService: ICapabilitiesService,
+		@IQuickInputService private _quickInputService: IQuickInputService,
+		@IExtensionService private _extensionService: IExtensionService
+	) {
+		this._providerRegisterEvent = this._onProviderRegister.event;
+	}
+
+	/**
+	 * Runs the actions using the provider that supports the fileFormat provided.
+	 * @param fileFormat: fileformat of the underlying execution plan file. It is used to get the provider that support it.
+	 * @param action: executionPlanService action to be performed.
+	 */
+	private async _runAction<T>(fileFormat: string, action: (handler: azdata.executionPlan.ExecutionPlanServiceProvider) => Thenable<T>): Promise<T> {
+		let providers = Object.keys(this._capabilitiesService.providers);
+		if (!providers) {
+			providers = await new Promise(resolve => {
+				this._capabilitiesService.onCapabilitiesRegistered(e => {
+					resolve(Object.keys(this._capabilitiesService.providers));
+				});
+			});
+		}
+
+		let epProviders: string[] = [];
+		for (let i = 0; i < providers.length; i++) {
+			const providerCapabilities = this._capabilitiesService.getCapabilities(providers[i]);
+			if (providerCapabilities.connection.supportedExecutionPlanFileExtensions?.includes(fileFormat)) {
+				epProviders.push(providers[i]);
+			}
+		}
+
+		let selectedProvider: string;
+		if (epProviders.length > 1) {
+			const providerQuickPick = this._quickInputService.createQuickPick<IQuickPickItem>();
```

Comment:
what's the scenario when end up here?  why can't we determine which execution plan provider that should be used (from extension or connection provider type or something)?  This may be a place where source code comments add some value.

Reply:
There are some providers like Postgres and MYSQL that allow you to FORMAT execution plan in different file formats. In those cases, there can be some conflicting file extensions that are supported by more than one provider. 
This handles that. 
https://www.postgresql.org/docs/9.1/sql-explain.html